### PR TITLE
expression: fix incorrect result of logical operators

### PR DESCRIPTION
**Pull request by sduzh**
### What problem does this PR solve? 
Fix issue #11199

```
TiDB> SELECT NULL OR 0.00000001000000000000;
+--------------------------------+
| NULL OR 0.00000001000000000000 |
+--------------------------------+
|                              1 |
+--------------------------------+
1 row in set (0.00 sec)

TiDB> SELECT 0 OR 0.00000001000000000000;
+-----------------------------+
| 0 OR 0.00000001000000000000 |
+-----------------------------+
|                           1 |
+-----------------------------+
1 row in set (0.00 sec)

TiDB> SELECT 0 OR "0.01a";
+--------------+
| 0 OR "0.01a" |
+--------------+
|            1 |
+--------------+
1 row in set, 1 warning (0.00 sec)
```

### What is changed and how it works?
Wrap the input parameters of the builtin `AND`、`OR` and `XOR` operators with `IS TRUE` function before cast them to `int` type.

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

Release note

### CODE REVIEW

**Comment by zz-jason** [comment]
@sduzh should we directly modify `builtinLogicOrSig` to return 1 when the two input parameters are `null` and `1`?

**Comment by sduzh** [comment]
> @sduzh should we directly modify `builtinLogicOrSig` to return 1 when the two input parameters are `null` and `1`?

@zz-jason Sorry, I didn't understand what you mean. Isn't this the current implementation?

File: expression/builtin_op.go [hunk 1]
```diff
@@ -377,7 +385,8 @@ func (b *builtinRightShiftSig) evalInt(row chunk.Row) (int64, bool, error) {
 
 type isTrueOrFalseFunctionClass struct {
 	baseFunctionClass
-	op opcode.Op
+	op       opcode.Op
+	castNull bool
```
Comment (qw4990) [inline expression/builtin_op.go]: Please add some comments for this field.

File: expression/expression.go [hunk 2]
```diff
@@ -407,6 +408,18 @@ func NewValuesFunc(ctx sessionctx.Context, offset int, retTp *types.FieldType) *
 	}
 }
 
+// NewIsTrueFunc create a new istrue function.
+func NewIsTrueFunc(ctx sessionctx.Context, castNull bool, arg Expression) *ScalarFunction {
```
Comment (SunRunAway) [inline expression/expression.go]: Does it not have to be exported?

File: expression/expression.go [hunk 3]
```diff
@@ -407,6 +408,18 @@ func NewValuesFunc(ctx sessionctx.Context, offset int, retTp *types.FieldType) *
...
+	fc := &isTrueOrFalseFunctionClass{baseFunctionClass{ast.IsTruth, 1, 1}, opcode.IsTruth, castNull}
+	f, err := fc.getFunction(ctx, []Expression{arg})
+	terror.Log(err)
```
Comment (SunRunAway) [inline expression/expression.go]: Should we throw this error out?

File: expression/expression.go [hunk 4]
```diff
@@ -407,6 +408,18 @@ func NewValuesFunc(ctx sessionctx.Context, offset int, retTp *types.FieldType) *
 	}
 }
 
+// NewIsTrueFunc create a new istrue function.
```
Comment (SunRunAway) [inline expression/expression.go]: nit: Could you add a comment here to explain why this function should be present and what does `castNull` mean?

File: expression/builtin_op.go (hunk 1 above)
Comment (SunRunAway) [reply expression/builtin_op.go]: Is it better to keep it false builtin.go? Would it be a better option to naming it `keepNull` ?

File: expression/expression.go (hunk 4 above)
Comment (sduzh) [reply expression/expression.go]: Would it be more elegant to add a new eval type `EBoolean` for internal usage and modify the `CAST` function to support the cast to boolean?

File: expression/builtin_op.go (hunk 1 above)
Comment (sduzh) [reply expression/builtin_op.go]: updated

File: expression/expression.go (hunk 3 above)
Comment (sduzh) [reply expression/expression.go]: updated

**Comment by sduzh** [comment]
The [Jenkins job](https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tidb_ghpr_integration_common_test/detail/tidb_ghpr_integration_common_test/5849/pipeline#step-331-log-98) failed on the `udf_logic_xor` test:
```
SELECT a, b, a xor b FROM (SELECT `pk` AS a, `col_decimal_6_3_signed` AS b FROM `table50_int_autoinc`) tmp WHERE (a xor b) = 1 ORDER BY LOWER(a), LOWER(b);
a	b	a xor b
...     ...    ...
14	0.112	1
```

I tried reproducing the error and was unable to. Following is my reproducing steps.
```
TiDB> create table table50_int_autoinc(pk int not null auto_increment primary key, col_decimal_6_3_signed decimal(6,3));
Query OK, 0 rows affected (0.01 sec)

TiDB> insert into table50_int_autoinc values(14, 0.112);
Query OK, 1 row affected (0.00 sec)

TiDB> SELECT a, b, a xor b FROM (SELECT `pk` AS a, `col_decimal_6_3_signed` AS b FROM `table50_int_autoinc`) tmp WHERE (a xor b) = 1 ORDER BY LOWER(a), LOWER(b);
Empty set (0.00 sec)
```

PTAL @zz-jason @SunRunAway

**Comment by SunRunAway** [comment]
@sduzh 

In MySQL5.7, consider this case

```mysql
mysql> select 1 xor 0.5;
+-----------+
| 1 xor 0.5 |
+-----------+
|         0 |
+-----------+
1 row in set (0.00 sec)

mysql> select 1 xor 0.49;
+------------+
| 1 xor 0.49 |
+------------+
|          1 |
+------------+
1 row in set (0.00 sec)
```

It seems `xor` rounds its arguments in MySQL 5.7.

**Comment by SunRunAway** [comment]
@sduzh Please help resolve the conflicting files.

**Pull request merged**